File: csharp/react/React.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Reactor
{
    public InputCell CreateInputCell(int value) => 
        new InputCell(value);

    public ComputeCell CreateComputeCell(IEnumerable<Cell> producers, Func<int[], int> compute) => 
        new ComputeCell(producers, compute);
}

public abstract class Cell
{
    public virtual int Value { get; set; }
    public EventHandler<int> Changed;
}

public class InputCell : Cell
{
    private int value;
    public override int Value 
    { 
        get { return value; } 
        set 
        { 
            if(value == this.value)
                return;
            this.value = value; 
            Changed?.Invoke(this, value); 
        } 
    }

    public InputCell(int value) => Value = value;
}

public class ComputeCell : Cell
{
    private readonly IEnumerable<Cell> inputs;
    private readonly Func<int[], int> compute;

    private int GetValue() => compute(inputs.Select(o => o.Value).ToArray());

    private int oldValue;

    public override int Value { get => GetValue(); set => throw new InvalidOperationException(); }

    public ComputeCell(IEnumerable<Cell> inputs, Func<int[], int> compute)
    {
        this.inputs = inputs;
        this.compute = compute;    

        oldValue = GetValue();    

        foreach(var cell in inputs)
            cell.Changed += (o, e) =>
            {
                var newValue = GetValue();
                if(newValue == oldValue)
                    return;
                oldValue = newValue;
                Changed?.Invoke(this, newValue);
            };
    }
}

File: fsharp/react/React.fs
module React

[<AbstractClass>]
type Cell() =
    abstract Value: int with get, set
    abstract Changed: IEvent<int>

type InputCell (initialValue: int) =
    inherit Cell()

    let mutable value = initialValue
    let changed = new Event<int>()
    override __.Changed = changed.Publish

    override __.Value
        with get() = value
        and set(newValue) = 
            if newValue <> value then 
                value <- newValue
                changed.Trigger newValue
                
type ComputeCell (sources: Cell list, computation: int list -> int) =
    inherit Cell()

    let compute () = computation (sources |> List.map (fun o -> o.Value))
    let mutable oldValue = compute ()
    let changed = new Event<int>()

    let update () = 
        let newValue = compute()
        if newValue <> oldValue then 
            oldValue <- newValue
            changed.Trigger newValue
    do
        sources |> List.iter (fun o -> o.Changed.Add (fun _ -> update ()))

    override __.Changed = changed.Publish

    override __.Value
        with get() = compute()
        and set(_) = failwith "Compute Cells cant be set"

type Reactor() =
    member __.createInputCell initialValue = new InputCell (initialValue)
    member __.createComputeCell sources computation = new ComputeCell (sources, computation)